lib/sync.js

```diff
@@ -481,17 +481,31 @@ SyncApi.prototype._processRoomEvents = function(room, stateEventList,
 };
 
 function retryTimeMsForAttempt(attempt) {
-    // 2,4,8,16,32,64,128,128,128,... seconds
-    // max 2^7 secs = 2.1 mins
-    return Math.pow(2, Math.min(attempt, 7)) * 1000;
+    // 2,4,8,16,32,32,32,32,... seconds
+    // max 2^5 secs = 32 secs
+    return Math.pow(2, Math.min(attempt, 5)) * 1000;
 }
 
 function startSyncingRetryTimer(client, attempt, fn) {
     client._syncingRetry = {};
     client._syncingRetry.fn = fn;
+    var newAttempt = attempt;
+    var timeBeforeWaitingMs = Date.now();
+    var timeToWaitMs = retryTimeMsForAttempt(attempt);
     client._syncingRetry.timeoutId = setTimeout(function() {
-        fn();
-    }, retryTimeMsForAttempt(attempt));
+        var timeAfterWaitingMs = Date.now();
+        var timeDeltaMs = timeAfterWaitingMs - timeBeforeWaitingMs;
+        if (timeDeltaMs > (2 * timeToWaitMs)) {
+            // we've waited more than twice what we were supposed to. Reset the
```

Comment:
The 2x factor is mainly to provide stronger guarantees that this will fire infrequently, given we do warning logs. It would be annoying to see them constantly in the logs just because the DOM liked to block one iteration of the event loop for 2s * 0.1 = 200ms, which you've seen longer before e.g. https://github.com/vector-im/vector-web/issues/438

Will add a note about why this could happen.